### DISCUSSION TITLE
How to traverse AST

### DISCUSSION THREAD

**Discussion by spence-s** [harttle/liquidjs#294] [Q&A]
I would like to use the AST generated by liquidjs to create a liquid formatter. 

However, I am having trouble traversing the generated AST. 

```liquid
<div>
{%- assign color_brightness = color | color_brightness -%}

{%- if color_brightness < 26 -%}
  {%- if is_background_color -%}
    {%- assign new_color = color | color_lighten: 15 -%}
  {%- else -%}
    {%- assign new_color = color | color_lighten: 25 -%}
  {%- endif -%}
{%- endif -%}

</div>
{{- new_color -}}

<!-- /snippets/comment.liquid -->
<div class="comment__content rte">
  {{ comment.content }}
</div>
<div class="comment__meta">
  <span class="comment__meta-item">{{ comment.author }}</span>
  <span class="comment__meta-item">{{ comment.created_at | time_tag: format: 'date' }}</span>
</div>
```
A block like the above gets generated and I get an array of Nodes. However, I can't figure out how to get the tokens nested in the if statement. 

My generated output from looping through each Node and reducing getContent() or getText() to a string gives output like:
```liquid
<div>
{%- assign color_brightness = color | color_brightness -%}

{%- if color_brightness < 26 -%}
</div>
{{- new_color -}}
<!-- /snippets/comment.liquid -->
<div class="comment__content rte">
  
{{ comment.content }}

</div>
<div class="comment__meta">
  <span class="comment__meta-item">
{{ comment.author }}
</span>
  <span class="comment__meta-item">
{{ comment.created_at | time_tag: format: 'date' }}
</span>
</div>
```

Are there any suggestions or places in the code base you could point me to helping me traverse the AST better?

**Comment by spence-s** [comment]
I think I found the solution I need by running 

```js
 const tokenizer = new Tokenizer(text);
 const tokens = tokenizer.readTopLevelTokens();
```

Does this seem like the best method to reproduce the input liquid code from the source liquid code?

**Comment by harttle** [comment]
> Does this seem like the best method to reproduce the input liquid code from the source liquid code?

Yes, it is. And the result for tokenization is lexical AST (means there's no structural infomation). Calling `liquid.parse(html)` yields a syntaxical AST (also contains semantic infomation) but it's structure is not normalized and depends on implementation of tags.

> the best method to reproduce the input liquid code from the source liquid code

It's quite straightforward to reproduce the input, see the following code:
```javascript
const tokenizer = new Tokenizer(html)
const tokens = tokenizer.readTopLevelTokens()
const output = tokens.map(x => x.getText()).join('')
console.log(output === html)    // true
```

Note that not all subtypes of `Token` have `getContent()` method, but they all have `getText()`, see the [TypeDoc](https://liquidjs.com/api/classes/tokens_token_.token.html) for details.
